common/utils/demangle.hpp

```diff
@@ -0,0 +1,51 @@
+#ifndef KOKKOSTOOLS_COMMON_UTILS_DEMANGLE_HPP
+#define KOKKOSTOOLS_COMMON_UTILS_DEMANGLE_HPP
+
+#include <memory>
+#include <string>
+
+#if defined(__GXX_ABI_VERSION)
+#define HAVE_GCC_ABI_DEMANGLE
+#endif
+
+#if defined(HAVE_GCC_ABI_DEMANGLE)
+#include <cxxabi.h>
+#endif  // HAVE_GCC_ABI_DEMANGLE
+
+namespace KokkosTools
+{
+
+#if defined(HAVE_GCC_ABI_DEMANGLE)
+//! Demangle @p mangledName implementation.
+inline auto demangleNameImpl(const char* mangled_name)
+{
+  int status = 0;
+
+  std::unique_ptr<char, decltype(&std::free)> demangled_name(
+    abi::__cxa_demangle(mangled_name, nullptr, nullptr, &status),
+    &std::free
+  );
+
+  if(status != 0) demangled_name.reset();
+
+  return demangled_name;
+}
+#endif
+
+//! @overload
+inline std::string demangleName(const std::string& mangledName)
+{
+#if defined(HAVE_GCC_ABI_DEMANGLE)
+    auto demangled_name = demangleNameImpl(mangledName.c_str());
+
+    if(demangled_name != nullptr)
+    {
+        return std::string(demangled_name.get());
+    }
+#endif
+  return mangledName;
+}
```

Comment:
Is there any good reason not to combine these two functions given that they are very short and handling of the failure needs to be treated in two places this way?

```suggestion
inline std::string demangleName(const std::string& mangledName)
{
#if defined(HAVE_GCC_ABI_DEMANGLE)
   int status = 0;

    std::unique_ptr<char, decltype(&std::free)> demangled_name(
      abi::__cxa_demangle(mangled_name.c_str(), nullptr, nullptr, &status),
      &std::free
    );

    if(status == 0) return std::string(demangled_name.get());
    else demangled_name.reset();
#endif
  return mangledName;
}
```

Reply:
OK, better indeed :heavy_check_mark: